### PR TITLE
gh-109191: Fix build with newer editline

### LINKED ISSUE #109191: Build fails with recent libedit versions
# Bug report

### Bug description:

`readline_set_completion_display_matches_hook_impl` will use either `VFunction` (if `_RL_FUNCTION_TYPEDEF` is defined) or `rl_compdisp_func_t`. The former is deprecated but available in current readline, while the latter has historically not been available in libedit. However, libedit recently added `rl_compdisp_func_t` and removed `VFunction` entirely: http://cvsweb.netbsd.org/bsdweb.cgi/src/lib/libedit/readline/readline.h.diff?r1=1.53&r2=1.54&sortby=date

We ship a recent libedit in MacPorts, so we see a compilation failure like this:
```
./Modules/readline.c:448:10: error: use of undeclared identifier 'VFunction'; did you mean 'function'?
        (VFunction *)on_completion_display_matches_hook : 0;
         ^~~~~~~~~
```

It's easy enough for us to patch around this, but I guess if compatibility with both current libedit and the older version shipped by Apple is desired, there needs to be another check for the availability of this type.

### CPython versions tested on:

3.12

### Operating systems tested on:

macOS

### Linked PRs
* gh-110239
* gh-110562
* gh-110575

### PR DESCRIPTION
**Pull request by Bo98**
Fixes compile with BSD libedit versions since April 2023.

Requires backport to 3.12 (should cleanly apply).
Requires backport to 3.11 and 3.10 using this manual cherry-pick: https://github.com/Bo98/cpython/commit/b0c571f6ad0f233ba9d0f24e32fc8db1081be247
I believe 3.9 and older don't technically support libedit except Apple's 2012 version, but a similar cherry-pick could be made if desired.

* Issue: gh-109191

### CODE REVIEW

**Comment by erlend-aasland** [comment]
> Requires backport to 3.12 (should cleanly apply). Requires backport to 3.11 and 3.10 using this manual cherry-pick: [Bo98@b0c571f](https://github.com/Bo98/cpython/commit/b0c571f6ad0f233ba9d0f24e32fc8db1081be247) I believe 3.9 and older don't technically support libedit except Apple's 2012 version, but a similar cherry-pick could be made if desired.

FYI, 3.10 and older is in security-fix-mode; I don't think this applies as a security fix. Backports to 3.12 and 3.11 are allowed for bug-fixes; is this a bugfix?

File: configure.ac
Comment (erlend-aasland) [inline configure.ac]: Style nit; please follow the [GNU Autoconf style guide](https://www.gnu.org/savannah-checkouts/gnu/autoconf/manual/autoconf-2.71/html_node/Coding-Style.html).

```suggestion
    AC_CHECK_TYPE([rl_compdisp_func_t],
                  [AC_DEFINE([HAVE_RL_COMPDISP_FUNC_T], [1],
                             [Define if readline supports rl_compdisp_func_t])],
                  [],
                  [readline_includes])
```

Even though configure.ac is coloured with a variety of styles, I prefer to stick with a clean style for all new code.

**Review by erlend-aasland — COMMENTED** [review]
@corona10: you've recently dived into the readline issues. Would you mind taking a look at this PR?

File: configure.ac
Comment (Bo98) [reply configure.ac]: Sorry, fixed now!

I indeed just copy-pasted the exact style of the readline checks before it.

**Comment by Bo98** [comment]
> Backports to 3.12 and 3.11 are allowed for bug-fixes; is this a bugfix?

I consider it so. It's certainly not a new feature.

Compile error fixes, including those adjusting for new OS and library versions, seem to have been backported before? I consider this the same category as those.

**Comment by corona10** [comment]
> @corona10: you've recently dived into the readline issues. Would you mind taking a look at this PR?

I will take a look by this weekend, I need time to investigate it :)

**Comment by erlend-aasland** [comment]
@corona10: we should have a conditional CI matrix for readline changes, similar to our OpenSSL CI matrix.

**Review by erlend-aasland — APPROVED** [review]
This looks good to me. I'll wait for Donghee's approval before merging.

**Comment by erlend-aasland** [comment]
Oh, BTW, can you produce a NEWS entry for this, @Bo98?

**Comment by corona10** [comment]
> @corona10: we should have a conditional CI matrix for readline changes, similar to our OpenSSL CI matrix.

Good idea

File: Misc/NEWS.d/next/Build/2023-10-05-11-46-20.gh-issue-109191.imUkVN.rst
Comment (erlend-aasland) [inline Misc/NEWS.d/next/Build/2023-10-05-11-46-20.gh-issue-109191.imUkVN.rst]:
```suggestion
Fix compile error when building with recent versions of libedit.
```

**Review by corona10 — COMMENTED** [review]
@Bo98  cc @erlend-aasland 

Since the change was introduced quite recent days
http://cvsweb.netbsd.org/bsdweb.cgi/src/lib/libedit/readline/readline.h.diff?r1=1.53&r2=1.54&sortby=date

I think that we have to support old versions(1.53) too.
WDYT?

**Comment by corona10** [comment]
@erlend-aasland 

> Requires backport to 3.11 and 3.10 using this manual cherry-pick: https://github.com/Bo98/cpython/commit/b0c571f6ad0f233ba9d0f24e32fc8db1081be247

About the backport patch, I am not sure if we handle this as the bug or not?
If the user users CPython <= 3.12, they should use libedit <= 1.53 not the latest version.
How did we deal with this issue for OpenSSL case?

**Comment by Bo98** [comment]
> Since the change was introduced quite recent days http://cvsweb.netbsd.org/bsdweb.cgi/src/lib/libedit/readline/readline.h.diff?r1=1.53&r2=1.54&sortby=date
> 
> I think that we have to support old versions(1.53) too.

This PR still supports older versions as it checks for the existence of `rl_compdisp_func_t`, which didn't exist prior to 1.54. So < 1.54 will still use the VFunction type cast.

This is evidenced by the compile working with Apple's libedit, which is based on version 1.33.

> About the backport patch, I am not sure if we handle this as the bug or not?
IMO, If the user uses CPython <= 3.12, they should use libedit <= 1.53 not the latest version.
How did we deal with similar issue for OpenSSL case?

Had a dig into this, and found the OpenSSL 3 changes to `_hashlib` (https://github.com/python/cpython/issues/84659):

* 3.9 got initial support for the new API from OpenSSL 3.0 alpha. 3.8 didn't receive a backport for the new API, but did receive a backport to satisfy the compiler and silence deprecation warnings.
* Final release of OpenSSL 3.0 came when 3.11 was the `main` branch, but the fixes for the final release were backported to 3.10 and 3.9.

So I guess the bar, at least in the OpenSSL case, is completely new API counts as a new feature which means no backport. But adjustments to fix compile issues and some other runtime breakages seem to be ok to backport.

The fixes for libedit in this PR are only compile-level adjustments - there is no runtime change as it's just a type cast adjustment. Both branches of the `#if defined` are identical otherwise.

**Comment by erlend-aasland** [comment]
> I think that we have to support old versions(1.53) too. WDYT?

It seems to be there is no problem with old versions; `AC_CHECK_TYPE` will check for the type def and the correct signature will be used.

> About the backport patch, I am not sure if we handle this as the bug or not?

Build failure => build bug. IMO, it is a bug fix.

> IMO, If the user uses CPython <= 3.12, they should use libedit <= 1.53 not the latest version.

Why?

If we are to restrict the upper version requirement for a third-party dependency, we must document that, and preferrably also implement compile time and runtime checks for that. I don't think any other dependency has an upper version limit.

> How did we deal with similar issue for OpenSSL case?

Which issue were you thinking of?

**Review by corona10 — APPROVED** [review]
@Bo98 @erlend-aasland 
After rereading this issue and PR, I noticed that there was a misunderstanding from myself.
LGTM and let's add the backport patch,

**Comment by erlend-aasland** [comment]
> Sorry, @Bo98 and @corona10, I could not cleanly backport this to `3.11` due to a conflict.

@Bo98, do you want to do the backport?

**Pull request merged**
